File: every-agent-web/src/sdk/frames.ts
/** 线上帧类型(架构 §3)。与 every-agent-contract Frames.java 对齐。 */

export const PROTOCOL_VERSION = 2;

/** 前端流消费进度回报事件(worker 级输入频道,payload={taskId, creditIndex})。 */
export const STREAM_ACK = 'stream.ack';

export interface HelloFrame {
  type: 'hello';
  ver: number;
  role: 'frontend' | 'worker' | 'agent';
  apiKey: string;
  clientId: string;
  /** hub 连接凭证(保护 hub);旧协议/旧客户端可能不带,故为可选。 */
  hubKey?: string;
}

export interface WelcomeFrame {
  type: 'welcome';
  ver: number;
  sessionId?: string;
  serverTs?: number;
}

export interface SubFrame {
  type: 'sub';
  channel: string;
}

export interface UnsubFrame {
  type: 'unsub';
  channel: string;
}

export interface PubFrame {
  type: 'pub';
  mid: string;
  channel: string;
  event: string;
  ts?: number;
  payload?: unknown;
}

/** hub 扇出的消息帧。 */
export interface MsgFrame {
  type: 'msg';
  channel: string;
  event: string;
  /**
   * 事件序号(仅 stream 频道携带;其余频道无此字段)。
   * worker 以<b>字符串</b>传输 64 位 Snowflake seq(> Number.MAX_SAFE_INTEGER,
   * JSON number 会丢精度,详见 contract Frames.wirePub);number 仅为兼容旧后端/测试保留。
   */
  seq?: number | string;
  ts?: number;
  from?: { clientId: string; role: string; sessionId?: string };
  /**
   * 数据包协议操作语义(架构 §6.2 新协议,缺省按 "replace" 处理):
   * - "replace":无该 seq → 新建数据包;有该 seq → 用新帧整体替换;
   * - "append":取出旧包,把 payload.text 追加到对应字段(thinking→payload.thinking,
   *   delta→payload.content,其他 event→payload.content)。
   *
   * 后端(DataPusher)把 operate 放在 ext.operate(与 ext.persist 同层,经 hub 原样透传);
   * 本顶层字段保留作兼容——读取方应优先取 ext.operate,缺失时回退本字段。
   */
  operate?: string;
  /** 扩展元数据(hub 原样透传;任务流帧含 target/operate/persist 等)。 */
  ext?: any;
  payload?: any;
}

export interface ErrorFrame {
  type: 'error';
  code: string;
  message?: string;
}

export type ClientFrame = HelloFrame | SubFrame | UnsubFrame | PubFrame;
export type ServerFrame = WelcomeFrame | MsgFrame | ErrorFrame;

/** rpc 载荷(worker cmd 频道上的 event="rpc")。 */
export interface RpcRequest {
  reqId: string;
  method: string;
  params?: Record<string, unknown>;
}

export interface RpcOk {
  reqId: string;
  result: any;
}

export interface RpcErr {
  reqId: string;
  code: string;
  message: string;
}

export interface RpcData {
  reqId: string;
  batch: any[];
  hasMore: boolean;
}

export interface RpcProgress {
  reqId: string;
  message: string;
  pct?: number;
}

export class RpcError extends Error {
  constructor(readonly code: string, message: string) {
    super(`[${code}] ${message}`);
  }
}

File: every-agent-contract/src/main/java/dev/everyagent/contract/frame/Frames.java
package dev.everyagent.contract.frame;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import dev.everyagent.contract.json.Json;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.node.ObjectNode;

import java.util.Map;

/**
 * 控制帧(架构 §3.1)。接收侧 must-ignore:未知字段一律忽略。
 */
public final class Frames {

    // type
    public static final String HELLO = "hello";
    public static final String WELCOME = "welcome";
    public static final String SUB = "sub";
    public static final String UNSUB = "unsub";
    public static final String PUB = "pub";
    public static final String MSG = "msg";
    public static final String ERROR = "error";

    // error codes(hub 连接级,架构 §13.4)
    public static final String E_NOT_AUTHENTICATED = "NOT_AUTHENTICATED";
    public static final String E_VERSION_MISMATCH = "VERSION_MISMATCH";
    public static final String E_ACL_DENIED = "ACL_DENIED";
    public static final String E_FRAME_TOO_LARGE = "FRAME_TOO_LARGE";
    public static final String E_RATE_LIMITED = "RATE_LIMITED";

    // hub 生成的 stream 频道订阅通知事件(架构演进:hub 感知前端在 stream 频道上的 sub/unsub,
    // 以 msg 帧向该 owner 下全部在线 worker 投递;payload={sessionId, taskId})。
    public static final String SUBSCRIBER_JOIN = "subscriber.join";
    public static final String SUBSCRIBER_LEAVE = "subscriber.leave";

    /** v2:输入改走 worker 级频道(u.K.worker.<id>.input,弃 per-task input)+ 存储按用户隔离。 */
    public static final int PROTOCOL_VERSION = 2;

    private Frames() {
    }

    // ---- 帧 records(用于解析;构造发送帧用下面的 wire* 帮助方法)----

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record Hello(Integer ver, String role, String apiKey, String clientId, String hubKey, Map<String, Object> meta) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record Welcome(Integer ver, String sessionId, long serverTs) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record Sub(String channel) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record Unsub(String channel) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record Pub(String mid, String channel, String event, Long seq, long ts, JsonNode payload, JsonNode ext) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record From(String clientId, String role) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record Msg(String channel, String event, Long seq, long ts, From from, JsonNode payload, JsonNode ext) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record ErrorFrame(String code, String detail) {
    }

    // ---- 发送帧构造 ----

    public static String wireHello(String role, String apiKey, String clientId, Map<String, Object> meta) {
        return wireHello(role, apiKey, clientId, null, meta);
    }

    /** hello 携带可选 hubKey(对 hub 的保护);hubKey 为空时不上线字段(向后兼容)。 */
    public static String wireHello(String role, String apiKey, String clientId, String hubKey, Map<String, Object> meta) {
        ObjectNode o = Json.MAPPER.createObjectNode();
        o.put("type", HELLO);
        o.put("ver", PROTOCOL_VERSION);
        o.put("role", role);
        o.put("apiKey", apiKey);
        o.put("clientId", clientId);
        if (hubKey != null && !hubKey.isBlank()) {
            o.put("hubKey", hubKey);
        }
        if (meta != null && !meta.isEmpty()) {
            o.set("meta", Json.toJson(meta));
        }
        return Json.write(o);
    }

    public static String wireSub(String channel) {
        ObjectNode o = Json.MAPPER.createObjectNode();
        o.put("type", SUB);
        o.put("channel", channel);
        return Json.write(o);
    }

    public static String wireUnsub(String channel) {
        ObjectNode o = Json.MAPPER.createObjectNode();
        o.put("type", UNSUB);
        o.put("channel", channel);
        return Json.write(o);
    }

    /**
     * 构造 pub 帧。seq 仅 stream 频道事件携带,其余频道传 null。
     *
     * <p>seq 在 wire 上<b>一律以字符串传输</b>({@code "17592…"}):seq 是 64 位 Snowflake
     * long(≈ 10^17 量级),远超 JS {@code Number.MAX_SAFE_INTEGER}(2^53 ≈ 9.007e15)。
     * 若按 JSON number 输出,浏览器端 {@code JSON.parse} 会丢精度——同毫秒相邻事件(如
     * {@code message} 与其 {@code tool.result})会被解析成同一个 double,前端按 seq 去重时
     * 把后到的事件误判为旧轮/重复而丢弃(工具结果即因此不可见)。字符串化后 hub 原样透传,
     * 前端 {@code compareSeq} 按位数优先的字符串序比较,等价数值序且不丢精度。
     */
    public static String wirePub(String mid, String channel, String event, Long seq, long ts, JsonNode payload, JsonNode ext) {
        ObjectNode o = Json.MAPPER.createObjectNode();
        o.put("type", PUB);
        if (mid != null) {
            o.put("mid", mid);
        }
        o.put("channel", channel);
        o.put("event", event);
        if (seq != null) {
            o.put("seq", seq.toString());
        }
        o.put("ts", ts);
        o.set("payload", payload == null ? Json.MAPPER.createObjectNode() : payload);
        if (ext != null && !ext.isNull()) {
            o.set("ext", ext);
        }
        return Json.write(o);
    }
}
